source/vir/src/builtins.rs

```diff
@@ -23,3 +28,26 @@ pub fn krate_add_builtins(no_span: &Span, krate: &mut KrateX) {
         DatatypeX { path, visibility, transparency, typ_params, variants, mode: Mode::Exec };
     krate.datatypes.push(Spanned::new(no_span.clone(), datatypex));
 }
+
+pub fn krate_add_array(no_span: &Span, krate: &mut KrateX) {
+    let path = crate::def::array_type();
+    let visibility = Visibility { owning_module: None, is_private: false };
+    let transparency = DatatypeTransparency::Never;
+
+    // Create a fake variant; it shouldn't matter, since transparency is Never.
+    let fields = Arc::new(vec![]);
+    let variant = ident_binder(&Arc::new("DummyArrayVariant".to_string()), &fields);
+    let variants = Arc::new(vec![variant]);
+
+    let bound = Arc::new(GenericBoundX::Traits(vec![]));
+    let is_strictly_positive = true;
+
+    let typ_params = Arc::new(vec![
+        (crate::def::array_param_ty(), bound.clone(), is_strictly_positive),
+        (crate::def::array_param_len(), bound, is_strictly_positive),
+    ]);
+
+    let datatypex =
+        DatatypeX { path, visibility, transparency, typ_params, variants, mode: Mode::Exec };
+    krate.datatypes.push(Spanned::new(no_span.clone(), datatypex));
+}
```

Comment:
The fact that we're making up e.g. a variant here seems a bit hackish to me. What is using a `Datatype` buying us? The ability to use the `typ_params` machinery?